File: src/test/java/swagger/pet/store/UserTest.java
package swagger.pet.store;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

import org.json.JSONObject;
import org.junit.Test;

import commons.CommonFunctions;
import io.restassured.response.Response;

public class UserTest {
    @Test
    public void postUser() throws IOException {
        String jsonBody = new String(Files.readAllBytes(Paths.get("src/test/resources/json/postUser.json")));

        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "application/json");
        headers.put("Content-Type", "application/json");
        Response response = CommonFunctions.performRequest("POST", "/user", jsonBody, headers, null, 200);     
        String emailValue = response.jsonPath().getString("email");
        assertEquals("The email value response its not the expected", "john@email.com", emailValue);

        // Negative Scenario
        CommonFunctions.performRequest("POST", "/user", "null", headers, null, 500);     
    }

    @Test
    public void createUserWithList() throws IOException {
        String jsonBody = new String(Files.readAllBytes(Paths.get("src/test/resources/json/postUser.json")));
        jsonBody = "[" + jsonBody + "]";

        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "application/xml");
        headers.put("Content-Type", "application/json");
        Response response = CommonFunctions.performRequest("POST", "/user/createWithList", jsonBody, headers, null, 200);    

        String usernameValue = response.xmlPath().getString("Users.item.username");
        assertEquals("The username value response its not the expected", "theUser", usernameValue);

        // Negative Scenario
        CommonFunctions.performRequest("POST", "/user/createWithList", "null", headers, null, 500);
    }

    @Test
    public void loginUser() {
        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "application/xml");
        Response response = CommonFunctions.performRequest("GET", "/user/login?username=string&password=string", null, headers, null, 200);     

        String[] parts = response.getBody().asString().split("Logged in user session: ");
        if (parts.length > 1) {
            String loggedInUserSession = parts[1].trim();
            assertNotNull("The response should contain a non-null loggedInUserSession", loggedInUserSession);
            assertTrue("The loggedInUserSession should be 17 characters long", loggedInUserSession.length() >= 17);
        } else {
            fail("The response does not contain the expected format");
        }
    }

    @Test
    public void logoutUser() {
        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "*/*");
        Response response = CommonFunctions.performRequest("GET", "/user/logout", null, headers, null, 200);  
        String responseBody = response.getBody().asString();
        assertTrue("The response does not contain 'User logged out'", responseBody.contains("User logged out"));        
    }

    @Test
    public void getUser() {
        String username = "theUser";
        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "application/xml");
        Response response = CommonFunctions.performRequest("GET", "/user/" + username, null, headers, null, 200);
        String usernameValue = response.xmlPath().getString("User.username");
        assertEquals("The username value response its not the expected", username, usernameValue);

        // Negative Scenario
        CommonFunctions.performRequest("GET", "/user/theUser2", null, headers, null, 404);  
    }

    @Test
    public void updateUser() throws IOException {
        String jsonBody = new String(Files.readAllBytes(Paths.get("src/test/resources/json/putUser.json")));
        
        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "*/*");
        headers.put("Content-Type", "application/json");
        Response response = CommonFunctions.performRequest("PUT", "/user/theUser", jsonBody, headers, null, 200);  
        String emailValue = response.jsonPath().getString("email");
        assertEquals("The email value response its not the expected", "john@email.com", emailValue);
        
        // Negative Scenario
        CommonFunctions.performRequest("PUT", "/user/theUser22", jsonBody, headers, null, 404); 
        CommonFunctions.performRequest("PUT", "/user/theUser", "null", headers, null, 500); 
    }

    @Test
    public void deleteUser() {
        Map<String, String> headers = new HashMap<>();
        headers.put("accept", "*/*");
        Response response = CommonFunctions.performRequest("DELETE", "/user/theUser", null, headers, null, 200);  

        // Create again the userID=11
        try {
            postUser();
        } catch (IOException e) {
            System.err.println("Error occurred while creating the user: " + e.getMessage());
            throw new RuntimeException("Failed to create user after deletion", e);
        }            
    }    
}

File: src/k6/UserTest.js
import { check, sleep } from 'k6';
import { performRequest } from './CommonFuction.js';
import { htmlReport } from "https://raw.githubusercontent.com/benc-uk/k6-reporter/main/dist/bundle.js";

const timeSleep = 1;

export function handleSummary(data) {
    return {
      "UserTest.html": htmlReport(data),
    };
  }

export const options = {
    scenarios: generateScenarios([
        'postUser',
        'createUserWithList',
        'loginUser',
        'logoutUser',
        'getUser',
        'updateUser',
        'deleteUser'
    ]),
};

function generateScenarios(functionNames) {
    let scenarios = {};
    functionNames.forEach((funcName) => {
        scenarios[`${funcName}Scenario`] = {
            executor: 'constant-vus',
            exec: funcName,
            vus: 1,
            duration: '1s',
        };
    });
    return scenarios;
}

export function postUser() {
    const jsonBody = JSON.stringify({
        "id": 11,
        "username": "theUser",
        "firstName": "John",
        "lastName": "James",
        "email": "john@email.com",
        "password": "12345",
        "phone": "12345",
        "userStatus": 1
      });

    const headers = {
        'accept': 'application/json',
        'Content-Type': 'application/json'
    };

    let res = performRequest('POST', '/user', jsonBody, headers, {}, 200);
    let emailValue = JSON.parse(res.body).email;
    check(res, {
        'email value is john@email.com in post User': (r) => emailValue === 'john@email.com',
    });

    // Negative Scenario
    performRequest('POST', '/user', 'null', headers, {}, 500);

    sleep(timeSleep);
}

export function createUserWithList() {
    const jsonBody = JSON.stringify([[{
        "id": 11,
        "username": "theUser",
        "firstName": "John",
        "lastName": "James",
        "email": "john@email.com",
        "password": "12345",
        "phone": "12345",
        "userStatus": 1
      }]]);

    const headers = {
        'accept': 'application/xml',
        'Content-Type': 'application/json'
    };

    let res = performRequest('POST', '/user/createWithList', jsonBody, headers, {}, 200);
    let usernameValue = res.body.match(/<username>([^<]+)<\/username>/)[1];
    check(res, {
        'username value is theUser in create User with List': (r) => usernameValue === 'theUser',
    });

    // Negative Scenario
    performRequest('POST', '/user/createWithList', 'null', headers, {}, 500);

    sleep(timeSleep);
}

export function loginUser() {
    const headers = {
        'accept': 'application/xml'
    };

    let res = performRequest('GET', '/user/login?username=string&password=string', null, headers, {}, 200);
    let loggedInUserSession = res.body.split('Logged in user session: ')[1]?.trim();
    
    check(res, {
        'loggedInUserSession is not null': (r) => loggedInUserSession !== undefined && loggedInUserSession !== null,
        'loggedInUserSession length is at least 17': (r) => loggedInUserSession && loggedInUserSession.length >= 17,
    });

    sleep(timeSleep);
}

export function logoutUser() {
    const headers = {
        'accept': '*/*'
    };

    let res = performRequest('GET', '/user/logout', null, headers, {}, 200);
    check(res, {
        'response contains User logged out': (r) => r.body.includes('User logged out'),
    });

    sleep(timeSleep);
}

export function getUser() {
    const username = 'theUser';
    const headers = {
        'accept': 'application/xml'
    };

    let res = performRequest('GET', `/user/${username}`, null, headers, {}, 200);
    let usernameValue = res.body.match(/<username>([^<]+)<\/username>/)[1];
    check(res, {
        'username value is theUser in get User': (r) => usernameValue === username,
    });

    // Negative Scenario
    performRequest('GET', '/user/theUser2', null, headers, {}, 404);

    sleep(timeSleep);
}

export function updateUser() {
    const jsonBody = JSON.stringify({
        "id": 10,
        "name": "doggie",
        "category": {
          "id": 1,
          "name": "Dogs"
        },
        "photoUrls": [
          "string"
        ],
        "tags": [
          {
            "id": 0,
            "name": "string"
          }
        ],
        "status": "available"
    });

    const headers = {
        'accept': '*/*',
        'Content-Type': 'application/json'
    };

    let res = performRequest('PUT', '/user/theUser', jsonBody, headers, {}, 200);
    let emailValue = JSON.parse(res.body).email;
    check(res, {
        'email value is john@email.com in update User': (r) => emailValue === 'john@email.com',
    });

    // Negative Scenarios
    performRequest('PUT', '/user/theUser22', jsonBody, headers, {}, 404);
    performRequest('PUT', '/user/theUser', 'null', headers, {}, 500);

    sleep(timeSleep);
}

export function deleteUser() {
    const headers = {
        'accept': '*/*'
    };

    let res = performRequest('DELETE', '/user/theUser', null, headers, {}, 200);
    check(res, {
        'status is 200': (r) => r.status === 200,
    });
    postUser();

    sleep(timeSleep);
}
